front/src/applications/operationalStudies/components/MacroEditor/utils.ts

```diff
@@ -244,29 +248,98 @@ export const getTrainrunFrequencyFromTimetableItem = (
-    color: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
-    colorFocus: CATEGORY_COLOR_VARIANTS[`${baseColor}70`],
-    colorMuted: CATEGORY_COLOR_VARIANTS[`${baseColor}10`],
-    colorRelated: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
-    colorDarkMode: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
-    colorDarkModeFocus: CATEGORY_COLOR_VARIANTS[`${baseColor}70`],
-    colorDarkModeMuted: CATEGORY_COLOR_VARIANTS[`${baseColor}10`],
-    colorDarkModeRelated: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
-  }));
+export const getNetzgrafikColors = async ({
+  categories,
+  dispatch,
+}: {
+  categories: TrainrunCategory[];
+  dispatch: AppDispatch;
+}) => {
+  const mainColors = Object.entries(NETZGRAFIK_COLOR_PALETTE).map(
+    ([colorRef, baseColor], index) => ({
+      id: index,
+      colorRef,
+      color: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
+      colorFocus: CATEGORY_COLOR_VARIANTS[`${baseColor}70`],
+      colorMuted: CATEGORY_COLOR_VARIANTS[`${baseColor}10`],
+      colorRelated: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
+      colorDarkMode: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
+      colorDarkModeFocus: CATEGORY_COLOR_VARIANTS[`${baseColor}70`],
+      colorDarkModeMuted: CATEGORY_COLOR_VARIANTS[`${baseColor}10`],
+      colorDarkModeRelated: CATEGORY_COLOR_VARIANTS[`${baseColor}50`],
+    })
+  );
+
+  const subCategoryPage = await dispatch(
+    osrdEditoastApi.endpoints.getSubCategory.initiate({ page: 1 }, { subscribe: false })
+  ).unwrap();
 
-export const getTrainMainCategoryFromId = (id: number): TrainMainCategory | null => {
-  for (const key of OSRD_TRAINRUN_CATEGORY_MAPPING.keys()) {
-    if (OSRD_TRAINRUN_CATEGORY_MAPPING.get(key)!.id === id) {
-      return key !== 'NO_CATEGORY' ? key : null;
+  const subCategories: SubCategory[] = [...subCategoryPage.results];
+
+  const subCategoriesFromTrainCategoryToNGE = categories.filter(
+    (_, i) => i > TRAIN_MAIN_CATEGORY_TO_NGE.length - 1
+  );
+
+  const subCategoriesFormatted = subCategoriesFromTrainCategoryToNGE.map((cat) => {
+    const subCategory = subCategories.find((subCat) => subCat.code === cat.colorRef);
+    if (!subCategory) {
+      throw new Error(`SubCategory with code ${cat.colorRef} not found.`);
     }
+    return {
+      ...subCategory,
+      colorRef: subCategory.code,
+      color: subCategory.color,
+      hovered_color: subCategory.hovered_color,
+      background_color: subCategory.background_color,
+    };
+  });
+
+  const subColors = subCategoriesFormatted.map((cat, index) => ({
+    id: mainColors.length + index,
+    colorRef: cat.colorRef,
+    color: cat.color,
+    colorFocus: cat.hovered_color,
+    colorMuted: cat.background_color,
+    colorRelated: cat.color,
+    colorDarkMode: cat.color,
+    colorDarkModeFocus: cat.hovered_color,
+    colorDarkModeMuted: cat.background_color,
+    colorDarkModeRelated: cat.color,
+  }));
+
+  return [...mainColors, ...subColors];
+};
+
+export const getTrainrunCategoryCodeFromId = (
```

Comment:
```suggestion
export const getTrainCategoryCodeFromTrainrunCategoryId = (
```

Nit: same as naming comments

Reply:
Done